modules/packages/OrderedSet/Treap.chpl

```diff
@@ -0,0 +1,862 @@
+
+      :returns: a tuple containing result
+      :rtype: `(bool, eltType)`
+    */
+    proc const upperBound(e: eltType): (bool, eltType) {
+      if !isDefaultInitializable(e) {
+        compilerError("upperBound is not avaliable on types that can't be \
+                      default-initialized, here: " + eltType: string);
+      }
+
+      _enter(); defer _leave();
+
+      var node = _upper_bound(_root, e);
+      var result: (bool, eltType);
+      if node == nil then  result[0] = false;
+      else result = (true, node!.element);
+      return result;
+    }
+
+    /*
+      Find the predecessor of one element in the orderedSet.
+      Returns if there is such one element.
+
+      Returns a tuple containing two element:
+      The first element is a `bool` that indicates whether there is such an element.
+      The second element is the occurence in the orderedSet, if there's any.
+
+      :arg e: The element to base
+      :type e: `eltType`
+
+      :returns: a tuple containing result
+      :rtype: `(bool, eltType)`
+    */
+    proc const predecessor(e: eltType, ref result: eltType): (bool, eltType) {
+      if !isDefaultInitializable(e) {
+        compilerError("predecessor is not avaliable on types that can't be \
+                      default-initialized, here: " + eltType: string);
+      }
+
+      _enter(); defer _leave();
+
+      var result: (bool, eltType);
+      result[0] = false;
+      var baseNode = _find(_root, e);
+      if baseNode == nil {
+        return result;
+      }
+
+      var resultNode = _neighbour(baseNode, 0);
+      if resultNode != nil {
+        result = (true, resultNode!.element);
+      }
+      return result;
+    }
+    
+    /*
+      Find the successor of one element in the orderedSet.
+      Returns if there is such one element.
+      If there is, store the result in `result`.
+
+      Returns a tuple containing two element:
+      The first element is a `bool` that indicates whether there is such an element.
+      The second element is the occurence in the orderedSet, if there's any.
+
+      :arg e: The element to base
+      :type e: `eltType`
+
+      :returns: a tuple containing result
+      :rtype: `(bool, eltType)`
+    */
+    proc const successor(e: eltType, ref result: eltType): (bool, eltType) {
+      if !isDefaultInitializable(e) {
+        compilerError("successor is not avaliable on types that can't be \
+                      default-initialized, here: " + eltType: string);
+      }
+
+      _enter(); defer _leave();
+
+      var result: (bool, eltType);
+      result[0] = false;
+      var baseNode = _find(_root, e);
+      if baseNode == nil {
+        return result;
+      }
+
+      var resultNode = _neighbour(baseNode, 1);
+      if resultNode != nil {
+        result = (true, resultNode!.element);
+      }
+      return result;
+    }
+
+    pragma "no doc"
+    proc _kth(node: nodeType, in k: int): nodeType {
```

Comment:
A line of comment will be helpful.

Reply:
I suppose it's a convention that each method has a lock-free verison whose name starts with a underline.

Reply:
Not really. We don't have such in-detail conventions as of today.

See: https://github.com/chapel-lang/chapel/issues/6698

p.s. not really reviewing the PR but +1 on Krishna's comment above

Reply:
https://github.com/chapel-lang/chapel/pull/16124/commits/26f4adca1da7cbc8b63f3e8a56f958b34288226e

Reply:
It's often a convention that `_foo()` will be a helper function for `foo()` in some way.  Having `foo()` acquire a lock and then call `_foo()` is common.  It can help with control flow in languages where you can't write `enter(); defer leave()`.  Or if the `foo()` operation is something you want to be able to call from other places that already hold the lock.

It can also be for things like if you wanted to have a public-facing `neighbor()` that takes the element value, and passes a module-private node reference to `_neighbor()`.  Or if the public `foo(n)` needs to pass an extra argument to a recursive `_foo(n, 0)`.

It eases the mental burden if you keep such a convention throughout a module or a project, that the underscore version always means the same thing, so you can certainly use it that way here.

Reply:
I've added a line of comment anyway.